Why is mypy so slow in Pyodide?

# Speeding up code

Do you know how to speed up the code? It takes less than a second to run on a standard Python installation without Pyodide, but over a minute with Pyodide:

```py
import os
os._exit = lambda arg: None

import mypy.main

with open("test.py", "w") as f:
    f.write('def test() -> None: return 5')

mypy.main.main(args=["test.py"])
```

Is Pyodide performance that bad?


# Timings

Running it twice in a row (using cache) still takes a while:
```py
# Pyodide: 60 seconds (creates cache)
# Local: < 1 second
mypy.main.main(args=["test.py"])

# Pyodide: 5 seconds (uses cache)
# Local: < 1 second
mypy.main.main(args=["test.py"])
```

Cache can be disabled, in which case, it takes the full time:
```py
# Pyodide: 60 seconds
# Local: < 1 second
mypy.main.main(args=["--cache-dir=/dev/null", "test.py"])

# Pyodide: 60 seconds
# Local: < 1 second
mypy.main.main(args=["--cache-dir=/dev/null", "test.py"])
```

_Originally posted by @ulmemxpoc in https://github.com/pyodide/pyodide/issues/3486#issuecomment-1399573992_

For me in Firefox 109.0 on Linux (pyodide 0.22.0, Py3.10), it takes 11s to run, while on host Python it's 800ms in my case. So ~10x slower. Looking at the profiling results, it looks like there are a lot of deeply nested function calls, which would be a somewhat unfavorable situation for benchmarks with Pyodide,
```py
>>> from pyinstrument import Profiler
profiler = Profiler()
>>> profiler.start(); mypy.main.main(args=["--cache-dir=/dev/null", "test.py"]); profiler.stop()
test.py:1: error: No return value expected  [return-value]
Found 1 error in 1 file (checked 1 source file)
<pyinstrument.session.Session object at 0x1b1b858>
>>> profiler.print()
```

<details>

```
>>> profiler.print(show_all=True)
  _     ._   __/__   _ _  _  _ _/_   Recorded: 00:04:41  Samples:  11441
 /_//_/// /_\ / //_// / //_'/ //     Duration: 11.571    CPU time: 0.000
/   _/                      v4.4.0
Program: 
11.571 <module>  <console>:1
`- 11.568 main  mypy/main.py:43
   `- 11.558 run_build  mypy/main.py:146
      `- 11.558 build  mypy/build.py:148
         `- 11.558 _build  mypy/build.py:209
            `- 11.539 dispatch  mypy/build.py:2846
               |- 10.305 process_graph  mypy/build.py:3159
               |  `- 10.301 process_stale_scc  mypy/build.py:3370
               |     |- 5.787 semantic_analysis_for_scc  mypy/semanal_main.py:72
               |     |  |- 2.865 process_top_levels  mypy/semanal_main.py:168
               |     |  |  `- 2.862 semantic_analyze_target  mypy/semanal_main.py:310
               |     |  |     `- 2.858 SemanticAnalyzer.refresh_partial  mypy/semanal.py:560
               |     |  |        `- 2.798 SemanticAnalyzer.refresh_top_level  mypy/semanal.py:584
               |     |  |           `- 2.756 SemanticAnalyzer.accept  mypy/semanal.py:6045
               |     |  |              |- 1.637 ClassDef.accept  mypy/nodes.py:1142
               |     |  |              |  `- 1.633 SemanticAnalyzer.visit_class_def  mypy/semanal.py:1430
               |     |  |              |     `- 1.578 SemanticAnalyzer.analyze_class  mypy/semanal.py:1438
               |     |  |              |        |- 0.698 SemanticAnalyzer.analyze_class_body_common  mypy/semanal.py:1534
               |     |  |              |        |  `- 0.652 Block.accept  mypy/nodes.py:1217
               |     |  |              |        |     `- 0.651 SemanticAnalyzer.visit_block  mypy/semanal.py:4250
               |     |  |              |        |        `- 0.646 SemanticAnalyzer.accept  mypy/semanal.py:6045
               |     |  |              |        |           |- 0.329 AssignmentStmt.accept  mypy/nodes.py:1300
               |     |  |              |        |           |  `- 0.328 SemanticAnalyzer.visit_assignment_stmt  mypy/semanal.py:2520
               |     |  |              |        |           `- 0.127 IfStmt.accept  mypy/nodes.py:1467
               |     |  |              |        |              `- 0.127 SemanticAnalyzer.visit_if_stmt  mypy/semanal.py:4342
               |     |  |              |        |- 0.376 SemanticAnalyzer.clean_up_bases_and_infer_type_variables  mypy/semanal.py:1666
               |     |  |              |        |  `- 0.173 SemanticAnalyzer.analyze_type_expr  mypy/semanal.py:6080
               |     |  |              |        |     `- 0.136 IndexExpr.accept  mypy/nodes.py:1893
               |     |  |              |        |        `- 0.136 SemanticAnalyzer.visit_index_expr  mypy/semanal.py:4777
               |     |  |              |        `- 0.180 SemanticAnalyzer.analyze_base_classes  mypy/semanal.py:1920
               |     |  |              |           `- 0.171 SemanticAnalyzer.expr_to_analyzed_type  mypy/semanal.py:6051
               |     |  |              |              `- 0.141 SemanticAnalyzer.anal_type  mypy/semanal.py:6123
               |     |  |              |                 `- 0.130 UnboundType.accept  mypy/types.py:815
               |     |  |              |                    `- 0.129 TypeAnalyser.visit_unbound_type  mypy/typeanal.py:235
               |     |  |              |                       `- 0.128 TypeAnalyser.visit_unbound_type_nonoptional  mypy/typeanal.py:243
               |     |  |              |- 0.737 AssignmentStmt.accept  mypy/nodes.py:1300
               |     |  |              |  `- 0.736 SemanticAnalyzer.visit_assignment_stmt  mypy/semanal.py:2520
               |     |  |              |     `- 0.295 SemanticAnalyzer.check_and_set_up_type_alias  mypy/semanal.py:3220
               |     |  |              |        `- 0.148 SemanticAnalyzer.analyze_alias  mypy/semanal.py:3175
               |     |  |              |           `- 0.136 analyze_type_alias  mypy/typeanal.py:121
               |     |  |              `- 0.308 IfStmt.accept  mypy/nodes.py:1467
               |     |  |                 `- 0.308 SemanticAnalyzer.visit_if_stmt  mypy/semanal.py:4342
               |     |  |                    `- 0.216 SemanticAnalyzer.visit_block  mypy/semanal.py:4250
               |     |  |                       `- 0.215 SemanticAnalyzer.accept  mypy/semanal.py:6045
               |     |  |- 2.283 process_functions  mypy/semanal_main.py:226
               |     |  |  `- 2.247 process_top_level_function  mypy/semanal_main.py:248
               |     |  |     `- 2.241 semantic_analyze_target  mypy/semanal_main.py:310
               |     |  |        `- 2.202 SemanticAnalyzer.refresh_partial  mypy/semanal.py:560
               |     |  |           |- 2.040 SemanticAnalyzer.accept  mypy/semanal.py:6045
               |     |  |           |  |- 1.511 FuncDef.accept  mypy/nodes.py:821
               |     |  |           |  |  `- 1.510 SemanticAnalyzer.visit_func_def  mypy/semanal.py:782
               |     |  |           |  |     `- 1.470 SemanticAnalyzer.analyze_func_def  mypy/semanal.py:810
               |     |  |           |  |        |- 0.591 TypeAnalyser.visit_callable_type  mypy/typeanal.py:850
               |     |  |           |  |        |  |- 0.192 TypeAnalyser.anal_array  mypy/typeanal.py:1386
               |     |  |           |  |        |  |  `- 0.169 TypeAnalyser.anal_type  mypy/typeanal.py:1394
               |     |  |           |  |        |  |- 0.144 CallableType.bind_function_type_variables  mypy/typeanal.py:1348
               |     |  |           |  |        |  |  `- 0.135 TypeAnalyser.infer_type_variables  mypy/typeanal.py:1325
               |     |  |           |  |        |  `- 0.131 TypeAnalyser.anal_type  mypy/typeanal.py:1394
               |     |  |           |  |        |     `- 0.120 UnboundType.accept  mypy/types.py:815
               |     |  |           |  |        |        `- 0.118 TypeAnalyser.visit_unbound_type  mypy/typeanal.py:235
               |     |  |           |  |        |           `- 0.116 TypeAnalyser.visit_unbound_type_nonoptional  mypy/typeanal.py:243
               |     |  |           |  |        |- 0.245 SemanticAnalyzer.analyze_function_body  mypy/semanal.py:1241
               |     |  |           |  |        `- 0.209 SemanticAnalyzer.update_function_type_variables  mypy/semanal.py:957
               |     |  |           |  |           `- 0.168 FuncDef.bind_function_type_variables  mypy/typeanal.py:1348
               |     |  |           |  |              `- 0.157 TypeAnalyser.infer_type_variables  mypy/typeanal.py:1325
               |     |  |           |  `- 0.527 OverloadedFuncDef.accept  mypy/nodes.py:616
               |     |  |           |     `- 0.527 SemanticAnalyzer.visit_overloaded_func_def  mypy/semanal.py:967
               |     |  |           |        `- 0.520 SemanticAnalyzer.analyze_overloaded_func_def  mypy/semanal.py:981
               |     |  |           |           |- 0.362 SemanticAnalyzer.analyze_overload_sigs_and_impl  mypy/semanal.py:1054
               |     |  |           |           |  `- 0.359 Decorator.accept  mypy/nodes.py:919
               |     |  |           |           |     `- 0.359 SemanticAnalyzer.visit_decorator  mypy/semanal.py:1346
               |     |  |           |           |        `- 0.344 FuncDef.accept  mypy/nodes.py:821
               |     |  |           |           |           `- 0.342 SemanticAnalyzer.visit_func_def  mypy/semanal.py:782
               |     |  |           |           |              `- 0.335 SemanticAnalyzer.analyze_func_def  mypy/semanal.py:810
               |     |  |           |           |                 `- 0.145 TypeAnalyser.visit_callable_type  mypy/typeanal.py:850
               |     |  |           |           `- 0.152 Decorator.accept  mypy/nodes.py:919
               |     |  |           |              `- 0.152 SemanticAnalyzer.visit_decorator  mypy/semanal.py:1346
               |     |  |           |                 `- 0.147 FuncDef.accept  mypy/nodes.py:821
               |     |  |           |                    `- 0.147 SemanticAnalyzer.visit_func_def  mypy/semanal.py:782
               |     |  |           |                       `- 0.147 SemanticAnalyzer.analyze_func_def  mypy/semanal.py:810
               |     |  |           `- 0.133 _GeneratorContextManager.__enter__  contextlib.py:130
               |     |  |              `- 0.132 SemanticAnalyzer.file_context  mypy/semanal.py:729
               |     |  |- 0.459 check_type_arguments  mypy/semanal_main.py:366
               |     |  |  `- 0.453 MypyFile.accept  mypy/nodes.py:372
               |     |  |     `- 0.453 TypeArgumentAnalyzer.visit_mypy_file  mypy/semanal_typeargs.py:48
               |     |  |        `- 0.453 TypeArgumentAnalyzer.visit_mypy_file  mypy/traverser.py:112
               |     |  |           `- 0.346 ClassDef.accept  mypy/nodes.py:1142
               |     |  |              `- 0.346 TypeArgumentAnalyzer.visit_class_def  mypy/semanal_typeargs.py:59
               |     |  |                 `- 0.339 TypeArgumentAnalyzer.visit_class_def  mypy/mixedtraverser.py:37
               |     |  |                    `- 0.327 TypeArgumentAnalyzer.visit_class_def  mypy/traverser.py:141
               |     |  |                       `- 0.325 Block.accept  mypy/nodes.py:1217
               |     |  |                          `- 0.325 TypeArgumentAnalyzer.visit_block  mypy/semanal_typeargs.py:63
               |     |  |                             `- 0.325 TypeArgumentAnalyzer.visit_block  mypy/traverser.py:116
               |     |  |                                `- 0.118 FuncDef.accept  mypy/nodes.py:821
               |     |  |                                   `- 0.118 TypeArgumentAnalyzer.visit_func_def  mypy/traverser.py:132
               |     |  |                                      `- 0.117 TypeArgumentAnalyzer.visit_func  mypy/semanal_typeargs.py:53
               |     |  `- 0.136 calculate_class_properties  mypy/semanal_main.py:461
               |     |- 3.984 State.type_check_first_pass  mypy/build.py:2304
               |     |  `- 3.971 TypeChecker.check_first_pass  mypy/checker.py:440
               |     |     `- 3.950 TypeChecker.accept  mypy/checker.py:570
               |     |        |- 2.998 ClassDef.accept  mypy/nodes.py:1142
               |     |        |  `- 2.998 TypeChecker.visit_class_def  mypy/checker.py:2124
               |     |        |     |- 2.420 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |  `- 2.420 Block.accept  mypy/nodes.py:1217
               |     |        |     |     `- 2.420 TypeChecker.visit_block  mypy/checker.py:2560
               |     |        |     |        `- 2.410 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |           |- 0.956 FuncDef.accept  mypy/nodes.py:821
               |     |        |     |           |  `- 0.955 TypeChecker.visit_func_def  mypy/checker.py:934
               |     |        |     |           |     `- 0.942 TypeChecker._visit_func_def  mypy/checker.py:940
               |     |        |     |           |        |- 0.636 TypeChecker.check_func_item  mypy/checker.py:995
               |     |        |     |           |        |  `- 0.568 TypeChecker.check_func_def  mypy/checker.py:1031
               |     |        |     |           |        |     `- 0.134 TypeChecker.check_reverse_op_method  mypy/checker.py:1442
               |     |        |     |           |        `- 0.280 TypeChecker.check_method_override  mypy/checker.py:1758
               |     |        |     |           |           `- 0.278 TypeChecker.check_method_or_accessor_override_for_base  mypy/checker.py:1769
               |     |        |     |           |              `- 0.262 TypeChecker.check_method_override_for_base_with_name  mypy/checker.py:1804
               |     |        |     |           |                 `- 0.128 TypeChecker.check_override  mypy/checker.py:1965
               |     |        |     |           |- 0.876 IfStmt.accept  mypy/nodes.py:1467
               |     |        |     |           |  `- 0.876 TypeChecker.visit_if_stmt  mypy/checker.py:4163
               |     |        |     |           |     |- 0.475 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |           |     |  `- 0.474 Block.accept  mypy/nodes.py:1217
               |     |        |     |           |     |     `- 0.474 TypeChecker.visit_block  mypy/checker.py:2560
               |     |        |     |           |     |        `- 0.473 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |           |     |           |- 0.265 OverloadedFuncDef.accept  mypy/nodes.py:616
               |     |        |     |           |     |           |  `- 0.265 TypeChecker.visit_overloaded_func_def  mypy/checker.py:606
               |     |        |     |           |     |           |     `- 0.265 TypeChecker._visit_overloaded_func_def  mypy/checker.py:612
               |     |        |     |           |     |           |        `- 0.258 TypeChecker.check_func_item  mypy/checker.py:995
               |     |        |     |           |     |           |           `- 0.258 TypeChecker.check_func_def  mypy/checker.py:1031
               |     |        |     |           |     |           |              `- 0.232 TypeChecker.expand_typevars  mypy/checker.py:1732
               |     |        |     |           |     |           |                 `- 0.124 expand_type  mypy/expandtype.py:51
               |     |        |     |           |     |           |                    `- 0.124 CallableType.accept  mypy/types.py:1791
               |     |        |     |           |     |           |                       `- 0.124 ExpandTypeVisitor.visit_callable_type  mypy/expandtype.py:222
               |     |        |     |           |     |           |                          `- 0.124 ExpandTypeVisitor.expand_types  mypy/expandtype.py:351
               |     |        |     |           |     |           |                             `- 0.121 UnionType.accept  mypy/types.py:2591
               |     |        |     |           |     |           |                                `- 0.121 ExpandTypeVisitor.visit_union_type  mypy/expandtype.py:329
               |     |        |     |           |     |           |                                   `- 0.116 make_simplified_union  mypy/typeops.py:423
               |     |        |     |           |     |           `- 0.141 FuncDef.accept  mypy/nodes.py:821
               |     |        |     |           |     |              `- 0.141 TypeChecker.visit_func_def  mypy/checker.py:934
               |     |        |     |           |     |                 `- 0.139 TypeChecker._visit_func_def  mypy/checker.py:940
               |     |        |     |           |     |                    `- 0.133 TypeChecker.check_func_item  mypy/checker.py:995
               |     |        |     |           |     |                       `- 0.131 TypeChecker.check_func_def  mypy/checker.py:1031
               |     |        |     |           |     `- 0.370 ExpressionChecker.accept  mypy/checkexpr.py:4638
               |     |        |     |           |        `- 0.362 ComparisonExpr.accept  mypy/nodes.py:1983
               |     |        |     |           |           `- 0.362 ExpressionChecker.visit_comparison_expr  mypy/checkexpr.py:2798
               |     |        |     |           |              `- 0.344 ExpressionChecker.check_op  mypy/checkexpr.py:3270
               |     |        |     |           |                 `- 0.270 ComparisonExpr.check_op_reversible  mypy/checkexpr.py:3093
               |     |        |     |           |                    `- 0.145 lookup_operator  mypy/checkexpr.py:3102
               |     |        |     |           |                       `- 0.132 analyze_member_access  mypy/checkmember.py:139
               |     |        |     |           |                          `- 0.125 _analyze_member_access  mypy/checkmember.py:203
               |     |        |     |           |- 0.278 OverloadedFuncDef.accept  mypy/nodes.py:616
               |     |        |     |           |  `- 0.278 TypeChecker.visit_overloaded_func_def  mypy/checker.py:606
               |     |        |     |           |     `- 0.278 TypeChecker._visit_overloaded_func_def  mypy/checker.py:612
               |     |        |     |           |        `- 0.190 TypeChecker.check_func_item  mypy/checker.py:995
               |     |        |     |           |           `- 0.181 TypeChecker.check_func_def  mypy/checker.py:1031
               |     |        |     |           `- 0.223 Decorator.accept  mypy/nodes.py:919
               |     |        |     |              `- 0.223 TypeChecker.visit_decorator  mypy/checker.py:4524
               |     |        |     |                 `- 0.179 TypeChecker.check_func_item  mypy/checker.py:995
               |     |        |     |                    `- 0.154 TypeChecker.check_func_def  mypy/checker.py:1031
               |     |        |     `- 0.295 TypeChecker.check_init_subclass  mypy/checker.py:2203
               |     |        |        `- 0.282 ExpressionChecker.accept  mypy/checkexpr.py:4638
               |     |        |           `- 0.279 ExpressionChecker.visit_call_expr  mypy/checkexpr.py:401
               |     |        |              `- 0.278 CallExpr.visit_call_expr_inner  mypy/checkexpr.py:422
               |     |        |                 `- 0.192 ExpressionChecker.accept  mypy/checkexpr.py:4638
               |     |        |                    `- 0.191 MemberExpr.accept  mypy/nodes.py:1766
               |     |        |                       `- 0.191 ExpressionChecker.visit_member_expr  mypy/checkexpr.py:2650
               |     |        |                          `- 0.182 ExpressionChecker.analyze_ordinary_member_access  mypy/checkexpr.py:2656
               |     |        |- 0.584 IfStmt.accept  mypy/nodes.py:1467
               |     |        |  `- 0.584 TypeChecker.visit_if_stmt  mypy/checker.py:4163
               |     |        |     |- 0.282 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |  `- 0.282 Block.accept  mypy/nodes.py:1217
               |     |        |     |     `- 0.282 TypeChecker.visit_block  mypy/checker.py:2560
               |     |        |     |        `- 0.279 TypeChecker.accept  mypy/checker.py:570
               |     |        |     |           `- 0.125 ClassDef.accept  mypy/nodes.py:1142
               |     |        |     |              `- 0.125 TypeChecker.visit_class_def  mypy/checker.py:2124
               |     |        |     `- 0.269 ExpressionChecker.accept  mypy/checkexpr.py:4638
               |     |        |        `- 0.245 ComparisonExpr.accept  mypy/nodes.py:1983
               |     |        |           `- 0.245 ExpressionChecker.visit_comparison_expr  mypy/checkexpr.py:2798
               |     |        |              `- 0.219 ExpressionChecker.check_op  mypy/checkexpr.py:3270
               |     |        |                 `- 0.186 ComparisonExpr.check_op_reversible  mypy/checkexpr.py:3093
               |     |        `- 0.245 AssignmentStmt.accept  mypy/nodes.py:1300
               |     |           `- 0.245 TypeChecker.visit_assignment_stmt  mypy/checker.py:2610
               |     |              `- 0.155 TypeChecker.check_type_alias_rvalue  mypy/checker.py:2662
               |     `- 0.525 State.finish_passes  mypy/build.py:2352
               |        `- 0.314 State._patch_indirect_dependencies  mypy/build.py:2406
               |           `- 0.312 TypeIndirectionVisitor.find_modules  mypy/indirection.py:27
               |              `- 0.312 TypeIndirectionVisitor._visit  mypy/indirection.py:31
               `- 1.234 load_graph  mypy/build.py:2997
                  `- 1.231 State.__init__  mypy/build.py:1892
                     `- 1.115 State.parse_file  mypy/build.py:2116
                        `- 1.037 BuildManager.parse_file  mypy/build.py:837
                           `- 1.036 parse  mypy/parse.py:8
                              `- 1.023 parse  mypy/fastparse.py:251
                                 `- 0.917 ASTConverter.visit  mypy/fastparse.py:442
                                    `- 0.917 ASTConverter.visit_Module  mypy/fastparse.py:841
                                       `- 0.902 ASTConverter.translate_stmt_list  mypy/fastparse.py:479
                                          `- 0.902 ASTConverter.visit  mypy/fastparse.py:442
                                             |- 0.573 ASTConverter.visit_ClassDef  mypy/fastparse.py:1094
                                             |  `- 0.550 ASTConverter.as_required_block  mypy/fastparse.py:576
                                             |     `- 0.526 ASTConverter.translate_stmt_list  mypy/fastparse.py:479
                                             |        `- 0.524 ASTConverter.visit  mypy/fastparse.py:442
                                             |           |- 0.380 ASTConverter.visit_FunctionDef  mypy/fastparse.py:857
                                             |           |  `- 0.380 ASTConverter.do_func_def  mypy/fastparse.py:865
                                             |           |     `- 0.117 ASTConverter.transform_args  mypy/fastparse.py:1017
                                             |           `- 0.123 ASTConverter.visit_If  mypy/fastparse.py:1201
                                             `- 0.209 ASTConverter.visit_If  mypy/fastparse.py:1201
                                                `- 0.117 ASTConverter.as_required_block  mypy/fastparse.py:576
```

</details>

In chrome 106, the performance was mostly identical in this case. 

Actually, the reason why mypy is faster on host python, is that they [ship binary wheel](https://pypi.org/project/mypy/#files) for those platforms (I imagine compiled with mypyc), while in Pyodide we are using the pure Python version. So it's not a very fair comparison :)

--- a/packages/mypy/meta.yaml
+++ b/packages/mypy/meta.yaml
@@ -8,7 +8,8 @@ source:
   url: https://files.pythonhosted.org/packages/0e/5c/fbe112ca73d4c6a9e65336f48099c60800514d8949b4129c093a84a28dc8/mypy-0.991.tar.gz
   sha256: 3c0165ba8f354a6d9881809ef29f1a9318a236a6d81c690094c5df32107bde06
 build:
-  script: export MYPY_USE_MYPYC=1
+  script: |
+    export MYPY_USE_MYPYC=1
     export MYPYC_OPT_LEVEL=3
 about:
   home: http://www.mypy-lang.org/

--- a/packages/mypy/meta.yaml
+++ b/packages/mypy/meta.yaml
@@ -1,0 +1,17 @@
+package:
+  name: mypy
+  version: "0.991"
+  top-level:
+    - mypyc
+    - mypy
+source:
+  url: https://files.pythonhosted.org/packages/0e/5c/fbe112ca73d4c6a9e65336f48099c60800514d8949b4129c093a84a28dc8/mypy-0.991.tar.gz
+  sha256: 3c0165ba8f354a6d9881809ef29f1a9318a236a6d81c690094c5df32107bde06
+build:
+  script: export MYPY_USE_MYPYC=1
+    export MYPYC_OPT_LEVEL=3
+about:
+  home: http://www.mypy-lang.org/
+  PyPI: https://pypi.org/project/mypy
+  summary: Optional static typing for Python
+  license: MIT License

